hbase-server/src/main/java/org/apache/hadoop/hbase/security/access/HDFSAclController.java

```diff
@@ -0,0 +1,634 @@
+
+  @Override
+  public Optional<MasterObserver> getMasterObserver() {
+    return Optional.of(this);
+  }
+
+  @Override
+  public void preMasterInitialization(final ObserverContext<MasterCoprocessorEnvironment> c)
+      throws IOException {
+    if (isHdfsAclEnabled(c.getEnvironment().getConfiguration())) {
+      MasterServices masterServices = null;
+      MasterCoprocessorEnvironment mEnv = c.getEnvironment();
+      if (mEnv instanceof HasMasterServices) {
+        masterServices = ((HasMasterServices) mEnv).getMasterServices();
+      }
+      if (masterServices == null) {
+        throw new RuntimeException("master services can not be null");
+      }
+      hdfsAclHelper = new HDFSAclHelper(masterServices);
+      pathHelper = hdfsAclHelper.getPathHelper();
+      fs = pathHelper.getFileSystem();
+      // Set public directory permission to 751 to make all users have access permission.
+      // And we also need the access permission of the parent of HBase root directory, but
+      // it's not set here, because the owner of HBase root directory may don't own permission
+      // to change it's parent permission to 751.
+      // The {root/.tmp} and {root/.tmp/data} directories are created to make global user HDFS
+      // acls can be inherited.
+      Path[] paths = new Path[] { pathHelper.getRootDir(), pathHelper.getDataDir(),
+          pathHelper.getTmpDir(), pathHelper.getTmpDataDir(), pathHelper.getArchiveDir(),
+          pathHelper.getArchiveDataDir(), pathHelper.getSnapshotRootDir() };
+      for (Path path : paths) {
+        if (!fs.exists(path)) {
+          fs.mkdirs(path);
+        }
+        fs.setPermission(path, ACL_ENABLE_PUBLIC_HFILE_PERMISSION);
+      }
+      // create snapshot restore directory
+      Path restoreDir = new Path(
+          mEnv.getConfiguration().get(SNAPSHOT_RESTORE_TMP_DIR, SNAPSHOT_RESTORE_TMP_DIR_DEFAULT));
+      if (!fs.exists(restoreDir)) {
+        fs.mkdirs(restoreDir);
+        fs.setPermission(restoreDir, ACL_ENABLE_RESTORE_HFILE_PERMISSION);
+      }
+    }
+  }
+
+  @Override
+  public void postStartMaster(ObserverContext<MasterCoprocessorEnvironment> ctx)
+      throws IOException {
+    try (Admin admin = ctx.getEnvironment().getConnection().getAdmin()) {
+      if (admin.tableExists(PermissionStorage.ACL_TABLE_NAME)) {
+        // check if hbase:acl table has 'm' CF
+        TableDescriptor tableDescriptor = admin.getDescriptor(PermissionStorage.ACL_TABLE_NAME);
+        boolean containHdfsAclFamily = Arrays.stream(tableDescriptor.getColumnFamilies())
+            .anyMatch(family -> Bytes.equals(family.getName(), HDFSAclStorage.HDFS_ACL_FAMILY));
+        if (!containHdfsAclFamily) {
+          TableDescriptorBuilder builder =
+              TableDescriptorBuilder.newBuilder(tableDescriptor).setColumnFamily(
+                ColumnFamilyDescriptorBuilder.newBuilder(HDFSAclStorage.HDFS_ACL_FAMILY).build());
+          admin.modifyTable(builder.build());
+        }
+      }
```

Comment:
What if the hbase:acl does not exist ? should throw an exception ?

Reply:
This coprocessor should be configured after the AccessController, if hbase:acl table does not exist, the AccessController will not work correctly firstly? 
Let me add some logs and throw an TableNotFoundException here.